Specifying additional docker network for Apache container

apparle opened the pull request:
This is a fix for https://github.com/nextcloud/all-in-one/discussions/5482

szaimen commented:
Hi, thanks a lot for your PR!

I am wondering if this would conflict with another workflow that we already have that also pushes regularly to gh-pages?
See https://github.com/nextcloud/all-in-one/blob/main/.github/workflows/helm-release.yml

szaimen commented on compose.yaml:
```diff
@@ -17,6 +17,7 @@ services:
       # AIO_COMMUNITY_CONTAINERS: # With this variable, you can add community containers very easily. See https://github.com/nextcloud/all-in-one/tree/main/community-containers#community-containers
       # APACHE_PORT: 11000 # Is needed when running behind a web server or reverse proxy (like Apache, Nginx, Caddy, Cloudflare Tunnel and else). See https://github.com/nextcloud/all-in-one/blob/main/reverse-proxy.md
       # APACHE_IP_BINDING: 127.0.0.1 # Should be set when running behind a web server or reverse proxy (like Apache, Nginx, Caddy, Cloudflare Tunnel and else) that is running on the same host. See https://github.com/nextcloud/all-in-one/blob/main/reverse-proxy.md
+      # APACHE_CONTAINER_ADDITIONAL_NETWORK: frontend_net # (Optional) Connect the apache container to an additional docker network. Needed when behind a web server or reverse proxy (like Apache, Nginx, Caddy, Cloudflare Tunnel and else) running in a different docker network on same server. See https://github.com/nextcloud/all-in-one/blob/main/reverse-proxy.md
```
I would probably name it like that
```suggestion
      # APACHE_ADDITIONAL_NETWORK: frontend_net # (Optional) Connect the apache container to an additional docker network. Needed when behind a web server or reverse proxy (like Apache, Nginx, Caddy, Cloudflare Tunnel and else) running in a different docker network on same server. See https://github.com/nextcloud/all-in-one/blob/main/reverse-proxy.md
```

szaimen commented on php/src/Data/ConfigurationManager.php:
```diff
@@ -888,6 +888,13 @@ public function DeleteCollaboraDictionaries() : void {
     }
 
+    public function GetApacheContainerAdditionalNetwork() : string {
+        $envVariableName = 'APACHE_CONTAINER_ADDITIONAL_NETWORK';
```
```suggestion
        $envVariableName = 'APACHE_ADDITIONAL_NETWORK';
```

szaimen commented on php/src/Docker/DockerActionManager.php:
```diff
@@ -897,11 +924,23 @@ public function ConnectMasterContainerToNetwork() : void
+        if ($container->GetIdentifier() === 'nextcloud-aio-apache') {
+            $apacheContainerAdditionalNetwork = $this->configurationManager->GetApacheContainerAdditionalNetwork();
+            if ($apacheContainerAdditionalNetwork !== '') {
+                $this->ConnectContainerIdToExistingNetwork($container->GetIdentifier(), $container->GetInternalPort(), $apacheContainerAdditionalNetwork);
```
Can we maybe simply re-use the ConnectContainerIdToNetwork function?

Possibly something like this? ConnectContainerIdToNetwork(nextcloud-aio-apache, $internalPort, $apacheContainerAdditionalNetwork)

apparle replied:
I intentionally separated it out because `ConnectContainerIdToNetwork()` also tries to create the network if it does not exist. 
When someone's using this feature, the network should be created externally and must already exist. If it does not exist, it is likely a typo or incorrect usage, and AIO should throw an error instead of silently creating a new network and proceeding.

If code duplication is a concern I can update the `ConnectContainerIdToNetwork()` to take an optional parameter `alsoCreateNetwork = true` with true as default value, and set to false when trying to connect to this feature.

szaimen replied:
I see but isn't it bad if the process fails? Will not people create addiitonal support requests in that case? I would probably rather go the way to also create the network if it does not exist instead of failing out with the last container...

szaimen commented on php/src/Docker/DockerActionManager.php:
```diff
@@ -844,6 +844,33 @@ private function DisconnectContainerFromBridgeNetwork(string $id) : void
         }
     }
 
+    private function ConnectContainerIdToExistingNetwork(string $id, string $internalPort, string $network) : void
```
```suggestion
```

szaimen commented on reverse-proxy.md:
```diff
@@ -44,9 +44,14 @@ All examples below will use port `11000` as `APACHE_PORT`. This port will be exp
+    For this setup, the reverse-proxy container and the nextcloud containers (specifically the `nextcloud-aio-mastercontainer` and `nextcloud-aio-apache`) need to be connected to the same bridge network. This can be achieved in one of these two ways:
 
-    Another option (actually the recommended way) in this case is to use `--network host` option (or `network_mode: host` for docker-compose) as setting for the reverse proxy container to connect it to the host network. If you are using a firewall on the server, you need to open ports 80 and 443 for the reverse proxy manually. By doing so, the default sample configurations that point at `localhos ...
+    * Reverse Proxy container is connected to `nextcloud-aio` network, either by setting it up under the same service (same docker compose file) or as a secondary external network in the reverse proxy container service.
+    * The front-facing nextcloud containers (`nextcloud-aio-mastercontainer` and `nextcloud-aio-apache`) are connected to the docker network of the reverse proxy container by specifying env variable `APACHE_CONTAINER_ADDITIONAL_NETWORK`.
+
+    After this the reverse-proxy can utilize Docker DNS networking to access nextcloud by container names at `http://nextcloud-aio-apache:$APACHE_PORT` and `https://nextcloud-aio-mastercontainer:8080`.
+
+    A third alternative is to utilize host networking instead of docker bridge networking: Specify `--network host` option (or `network_mode: host` for docker-compose) as setting for the reverse proxy container to connect it to the host network. If you are using a firewall on the server, you need to open ports 80 and 443 for the reverse proxy manually.  By doing so, the default sample configurati ...
```
I am not satisfied by the formatting here.

you start an enumeration but are not using that for all the options.

Also, it is not the right place to talk about the `https://nextcloud-aio-mastercontainer:8080` here. imho.

Addiitonally, this setting is only going to work if one uses `SKIP_DOMAIN_VALIDATION: true`. So this needs to be accounted for.

szaimen replied:
I am mentioning skip-domain-validation because people would first point at the domaincheck container and after the validation point at the apache container to make the validation work...

szaimen reviewed (changes requested):
Thank you for your PR! :)

I have some remarks below

szaimen commented:
Additionally, is it possible to add a check for this variable in start.sh so that only allowed characters are actually used?
See for example something like this but for strings: https://github.com/nextcloud/all-in-one/blob/af975f6fde8c81cdb9be6b95186c1301f719cbe4/Containers/mastercontainer/start.sh#L179-L188

apparle commented:
@szaimen I've made a bunch of changes incorporating your suggestions, added support for domaincheck container to make it all seamless and updated documentation. Can you take a look?

docjyJ reviewed (approved):
LGTM

szaimen commented on php/src/Docker/DockerActionManager.php:
```diff
@@ -844,44 +844,49 @@ private function DisconnectContainerFromBridgeNetwork(string $id) : void
-            if ($e->getCode() !== 409) {
-                throw new \Exception("Could not create the nextcloud-aio network: " . $e->getMessage());
-            }
+        if($createNetwork) {
```
```suggestion
        if ($createNetwork) {
```

szaimen commented on php/src/Docker/DockerActionManager.php:
```diff
@@ -844,44 +844,49 @@ private function DisconnectContainerFromBridgeNetwork(string $id) : void
         $url = $this->BuildApiUrl(
             sprintf('networks/%s/connect', $network)
         );
+        $json_payload = [ 'Container' => $id ];
+        if ($alias !== ''  ) {
+            $json_payload['EndpointConfig'] = [ 'Aliases' => [ $alias ] ];
```
```suggestion
        $jsonPayload = [ 'Container' => $id ];
        if ($alias !== ''  ) {
            $jsonPayload['EndpointConfig'] = ['Aliases' => [ $alias ]];
```

szaimen commented on php/src/Docker/DockerActionManager.php:
```diff
@@ -844,44 +844,49 @@ private function DisconnectContainerFromBridgeNetwork(string $id) : void
-                    'json' => [
-                        'container' => $id,
-                    ]
+                    'json' => $json_payload
```
```suggestion
                    'json' => $jsonPayload
```

szaimen commented on php/src/Docker/DockerActionManager.php:
```diff
@@ -897,11 +902,28 @@ public function ConnectMasterContainerToNetwork() : void
+
+        $apacheAdditionalNetwork = $this->configurationManager->GetApacheAdditionalNetwork();
+        if ($apacheAdditionalNetwork !== '') {
+            $this->ConnectContainerIdToNetwork('nextcloud-aio-mastercontainer', '', $apacheAdditionalNetwork, false);
```
Shouldnt this rather be done via the compose.yaml itself I am not sure why the mastercontainer should also be inside the additonal network? what is the benefit of that?

szaimen commented on php/src/Docker/DockerActionManager.php:
```diff
@@ -897,11 +902,28 @@ public function ConnectMasterContainerToNetwork() : void
+        if ($container->GetIdentifier() === 'nextcloud-aio-apache' || $container->GetIdentifier() === 'nextcloud-aio-domaincheck') {
+            $apacheAdditionalNetwork = $this->configurationManager->GetApacheAdditionalNetwork();
+            if ($apacheAdditionalNetwork !== '') {
+                $this->ConnectContainerIdToNetwork($container->GetIdentifier(), $container->GetInternalPort(), $apacheAdditionalNetwork, false, alias: $alias);
```
```suggestion
                $this->ConnectContainerIdToNetwork($container->GetIdentifier(), $container->GetInternalPort(), $apacheAdditionalNetwork, false, $alias);
```

szaimen reviewed (changes requested):
I have some additional suggestions.

Please also add your new env to https://github.com/nextcloud/all-in-one/blob/main/tests/QA/060-environmental-variables.md

szaimen commented on reverse-proxy.md:
```diff
@@ -44,9 +44,10 @@ All examples below will use port `11000` as `APACHE_PORT`. This port will be exp
-    Another option (actually the recommended way) in this case is to use `--network host` option (or `network_mode: host` for docker-compose) as setting for the reverse proxy container to connect it to the host network. If you are using a firewall on the server, you need to open ports 80 and 443 for the reverse proxy manually. By doing so, the default sample configurations that point at `localhos ...
+    The reverse-proxy container needs to be connected to the nextcloud containers. This can be achieved one of these 3 ways:
+    1. Utilize host networking instead of docker bridge networking: Specify `--network host` option (or `network_mode: host` for docker-compose) as setting for the reverse proxy container to connect it to the host network. If you are using a firewall on the server, you need to open ports 80 and 443 for the reverse proxy manually. With this setup, the default sample configurations with reverse-pro ...
+    1. Connect nextcloud's external-facing containers to the reverse-proxy's docker network by specifying env variable APACHE_ADDITIONAL_NETWORK. With this setup, the reverse proxy can utilize Docker bridge network's DNS name resolution to access nextcloud at `http://nextcloud-aio-apache:$APACHE_PORT`. Note, the specified network must already exist before Nextcloud AIO is started.
```
```suggestion
    1. Connect nextcloud's external-facing containers to the reverse-proxy's docker network by specifying env variable APACHE_ADDITIONAL_NETWORK. With this setup, the reverse proxy can utilize Docker bridge network's DNS name resolution to access nextcloud at `http://nextcloud-aio-apache:$APACHE_PORT`. ⚠️⚠️⚠️ Note, the specified network must already exist before Nextcloud AIO is started. Otherwise it will fail to start the container because the network is not existing.
```

szaimen commented on Containers/mastercontainer/start.sh:
```diff
@@ -193,6 +193,19 @@ It is set to '$APACHE_IP_BINDING'."
+It is set to '$APACHE_ADDITIONAL_NETWORK'."
+        exit 1
+    fi
+    tmp_out=$(sudo -u www-data docker network connect $APACHE_ADDITIONAL_NETWORK nextcloud-aio-mastercontainer 2>&1)
+    if [ -n "$tmp_out" ] && [[ ! "$tmp_out" =~ "nextcloud-aio-mastercontainer already exists in network $APACHE_ADDITIONAL_NETWORK" ]]; then
+        print_red "Unable to connect to $APACHE_ADDITIONAL_NETWORK. Cannot continue. Error: \n $tmp_out"
+        exit 1
+    fi
```
I am not sure about this. same question as above

szaimen replied:
@apparle can you please clarify?

apparle commented:
I replied inline but apparently my comments are all "pending" and I didn't realize they aren't posted for everyone. Clearly my interpretation of github UI is broken. Let me post comments directly, not inline.

apparle commented:
The intent is to allow for the master container to also be exposed via a reverse proxy. For example, I've my reverse proxy point `https://aio-nextcloud.mydomain.com` to `https://nextcloud-aio-mastercontainer:8080` to access the admin page.

The default compose.yml suggests `network_mode: bridge`, then internally connects the mastercontainer to `nextcloud-aio` and then later disconnects the mastercontainer from the bridge through cron job. Since we're manipulating the network, if a user is specifying an additional network for proxy setup through `APACHE_ADDITIONAL_NETWORK`, it's more seamless to have mastercontainer also be accessible through the same connection. 
Yes, the user can create this connection manually -- but then the user must specify not just the network for proxy but also create `nextcloud-aio` doing something like below, which is again not obvious for a naive user.
```
services:
  nextcloud-aio-mastercontainer:
    image: nextcloud/all-in-one:latest
...
    networks:
      - nextcloud-aio 
      - front_net

networks:
  front_net:
    external: true
  nextcloud-aio:
    name: nextcloud-aio
    driver: bridge
```
With the mastercontainer changes I've posed, I imagine the end-user flow to be:
1. User sets up some reverse proxy container with SSL in a docker container. Nginx Proxy Manager with let's encrypt + DuckDNS is extremely simple GUI based and amateur proof. There's at least 3-4 tutorials I've seen on internet which are straight forward to follow.
1. User sets up 2 different routes (just blindly):
    - Some aio URL (`https://aio-nextcloud.mydomain.com`) pointing to `https://nextcloud-aio-mastercontainer:8080` for the admin page.
    - Main nextcloud URL (`https://nextcloud.mydomain.com`) pointing to `http://nextcloud-aio-apache:11000`.
1. User launches the Nextcloud AIO with default compose file and specifies `APACHE_ADDITIONAL_NETWORK=<the network name of reverse proxy>`.
1. Follow all the steps in GUI, and because reverse proxy is already set up, everything including domaincheck just works.

(FWIW, I do intend to post this whole set up somewhere after this feature goes through.)

At a higher level, I suppose it's a software-architecture vs. user-experience question on how much do we do from within the mastercontainer for the sake of consistent user-experience & simplicity between both mastercontainer & apache container vs. leave it up to user to manipulate the compose file. Personally, I think this is a simple enough extension to mastercontainer with minimal downsides, and enables the super-simple reverse proxy + nextcloud setup that I described in above comment. And the `start.sh` change also has the nice side-benefit of doing an early error check with clear error message in red on the docker log, if the specified network has issues.

szaimen commented on Containers/mastercontainer/start.sh:
```diff
@@ -193,6 +193,19 @@ It is set to '$APACHE_IP_BINDING'."
+It is set to '$APACHE_ADDITIONAL_NETWORK'."
+        exit 1
+    fi
+    tmp_out=$(sudo -u www-data docker network connect $APACHE_ADDITIONAL_NETWORK nextcloud-aio-mastercontainer 2>&1)
+    if [ -n "$tmp_out" ] && [[ ! "$tmp_out" =~ "nextcloud-aio-mastercontainer already exists in network $APACHE_ADDITIONAL_NETWORK" ]]; then
+        print_red "Unable to connect to $APACHE_ADDITIONAL_NETWORK. Cannot continue. Error: \n $tmp_out"
+        exit 1
+    fi
```
so basically do this in a follow-up

szaimen commented on php/src/Docker/DockerActionManager.php:
```diff
@@ -897,11 +902,28 @@ public function ConnectMasterContainerToNetwork() : void
         // Don't disconnect here since it slows down the initial login by a lot. Is getting done during cron.sh instead.
         // $this->DisconnectContainerFromBridgeNetwork('nextcloud-aio-mastercontainer');
+
+        $apacheAdditionalNetwork = $this->configurationManager->GetApacheAdditionalNetwork();
+        if ($apacheAdditionalNetwork !== '') {
+            $this->ConnectContainerIdToNetwork('nextcloud-aio-mastercontainer', '', $apacheAdditionalNetwork, false);
+        }
```
and this too

apparle commented:
> Hi @apparle thanks for the detailed explanation!
> 
> I was wondering though if we could handle the mastercontainer networking change in a separate PR? (I'd like to merge this PR here without the mastercontainer network change and then do the mastercontainer stuff in a follow-up)
> 
> Is that fine for you?

Sounds good, let me make that change.

szaimen reviewed (approved):
LGTM, thanks!

szaimen commented:
This is now released with v9.9.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

ctag commented:
Just tested on :beta to fix a timeout issue setting up collabora. It worked for fixing the timeout in the settings page, but then I immediately ran into https://github.com/nextcloud/files_pdfviewer/issues/1080 apparently and no documents will open for editing to allow me to test the network/proxy stuff fully.
